rfcs/013-gitlab-spam-checker/013-gitlab-spam-checker.md

```diff
@@ -0,0 +1,93 @@
+
+GitLab has its [Spamcheck](https://docs.gitlab.com/ee/user/admin_area/reporting/spamcheck.html) service, which can be used to check for spam issues and snippets. This service does not check for spam in notes, projects, or user profiles. Currently, it is not possible to configure the service to delete spam automatically. Deploying GitLab's Spamcheck service also requires GitLab Enterprise Edition, which RenkuLab.io's GitLab is not currently running on.
+
+GitLab can make use of Akismet, a third-party-hosted AI spam filter. However, this only checks for spam on issues and notes, needs training, and would involve handing data over to a third party in the US.
+
+## Proposed Solution
+
+The proposed spam management service will integrate with GitLab via system hooks. When notes, projects, or users are created or updated, GitLab will send system hooks to a spam management service. System hooks request examples can be found [here](https://docs.gitlab.com/ee/administration/system_hooks.html#hooks-request-example).
+
+System hooks triggered on group, project, and user events can be triggered at the system level. Issues and comments at a project level would require a project-level system hook. This might be able to be configured when a project is added to KG. Project-level webhook events examples can be found [here](https://docs.gitlab.com/ee/user/project/integrations/webhook_events.html).
+
+Snippets are unable to trigger system hooks. Therefore a job triggered at regular intervals will be required to check for new snippets. This can be done by checking the snippet API for new snippets. Documentation on the snippet API can be found [here](https://docs.gitlab.com/ee/api/snippets.html).
+
+Ideally, the solution will also be able to check abuse reports manually submitted by users in GitLab and check the reported URL in the abuse report for spam content.
+
+A feedback mechanism would also be implemented, allowing the machine learning model to learn from mistakes and improve over time. This could be done by storing content in an S3 bucket and regularly re-training the model on a dataset including new data. The system should also keep a log of decisions made and the confidence level of the decision.
+
+Similarly to GitLab's Spamcheck service, storing the model in an S3 bucket and pulling it to the spam management service on start-up would allow the model to be updated without having to redeploy the spam management service.
+
+An allowlist of trusted email domain providers could be implemented to prevent false positives to ensure users with trusted email domains can post content without it being checked for spam. YAT already possesses a list of trusted email domains.
+
+This mermaid diagram provides an overview of how this system would work:
```

Comment:
I find the diagram a bit confusing....it looks like the spam checking service 'checks' different endpoints, but it's not clear what this check consists of...does it mean that it checks for new content/changes since the last check took place? Does it retrieve the content and aggregate it and send it to the classifier? Would it make sense to make the box containing the spam checking service larger such that it contains at least the model and perhaps a cleanup function and maybe some scheduler for checking issues which need to be proactively checked?

Reply:
I don't want to make too many assumptions about the design of the actual service yet so I'm trying to keep it quite high level. Think of the upper items as inputs/triggers to the service and the lower items as actions/outputs